include/native/pipeline.h

```diff
@@ -21,398 +21,514 @@
- void composed_pipeline(parallel_execution_native &p, InStream &qin, Stage const & s, OutStream &qout, std::vector<std::thread> & tasks){
-    tasks.push_back(
-      std::thread([&](){
-        using lambdaType = Stage;
-        p.register_thread();
+template <typename InQueue, typename Transformer, typename OutQueue>
+void composed_pipeline(parallel_execution_native & ex, InQueue & input_queue,
+                       Transformer && transform_op, OutQueue & output_queue, 
+                       std::vector<std::thread> & tasks)
+{
+  using namespace std;
+  tasks.push_back(
+    thread{[&](){
+      ex.register_thread();
         
-        auto item = qin.pop();
-        while(true){
-           if(!item) {
-                qout.push( typename OutStream::value_type::value_type () );
-                break;
-           }else{
-                auto out = typename OutStream::value_type::value_type( (*s)(item.value()) );
-                qout.push(out);
-           }
-           item = qin.pop();
+      auto item = input_queue.pop();
+      for (;;) {
+        using output_type = typename OutQueue::value_type;
+        using output_value_type = typename output_type::value_type;
+        if (!item) {
+          output_queue.push(output_value_type{}); 
+          break;
         }
-
-        p.deregister_thread();
+        else {
+          auto out = output_value_type{transform_op(item.value())};
+          output_queue.push(out);
+        }
+        item = input_queue.pop();
       }
-    ));
+
+      ex.deregister_thread();
+    }}
+  );
 }
 
 //Last stage
-template <typename Stream, typename Stage>
- void stages( parallel_execution_native & p,Stream& st, Stage && s ) {
-
-   p.register_thread();
-
-   typename Stream::value_type item;
-   std::vector<typename Stream::value_type> elements;
-   long current = 0;
-   if(p.ordering){
-     item = st.pop( );
-     while( item.first ) {
-        if(current == item.second){
-           s( item.first.value() );
-           current ++;
-        }else{
-           elements.push_back(item);
-        }
-        for(auto it = elements.begin(); it != elements.end(); it++){
-           if((*it).second == current) {
-              s((*it).first.value());
-              elements.erase(it);
-              current++;
-              break;
-           }
+template <typename InQueue, typename Consumer>
+void pipeline_impl(parallel_execution_native & ex, InQueue& input_queue, 
+                   Consumer && consume) 
+{
+  using namespace std;
+  using input_type = typename InQueue::value_type;
+
+  ex.register_thread();
+
+  vector<input_type> elements;
+  long current = 0;
+  if (ex.ordering){
+    auto item = input_queue.pop();
+    while (item.first) {
+      if(current == item.second){
+        consume(item.first.value());
+        current ++;
+      }
+      else {
+        elements.push_back(item);
+      }
+      // TODO: Probably find_if() + erase 
```

Comment:
Double check this:
~~~c++
auto it = find_if(elements.begin(), elements.end(), 
  [](auto x) { return x.second== current; });
if(it != elements.end()){
  consume(*it->first);
  elements.erase(it);
  current = it->second+1;
}
~~~